### PR TITLE
JavaScript Refactoring - Rename Identifier, Wrap selection, Convert to arrow function, Create Getters & Setters

### PR DESCRIPTION
**Pull request by navch**
This PR implements "rename" functionality in JS mode to enable intelligent rename feature by using Tern's scope analysis and inference functions. Renaming across multiple files has been intentionally blocked for the time being as we are not analyzing all the JS files in current project root right now.

The rename function works by selecting a variable def/ref or placing cursor in and using Ctrl+R.

This code is written by @swmitra and already reviewed by @petetnt ([PR 13047](https://github.com/adobe/brackets/pull/13047))

I just restructured this code and created new extension called JavascriptCodeRefactoring.

**Known issues:-**

In a particular case Tern inference engine is picking up wrong hit. For example :-
```javascript
function fn1(error) {
    error.a = "abc";
    var a = "test";
    console.log(a);
}
```
If we try to rename a in line 3 or 4 , Tern picks up fake ref from line 2 as well.

#Issue reported in tern [950](https://github.com/ternjs/tern/issues/950), [951](https://github.com/ternjs/tern/issues/951)
[951](https://github.com/ternjs/tern/issues/951) is already fixed.

**Wrap Selection in Try Catch/Condition** 
1) Select code/Just put cursor anywhere
2) Edit -> wrap in try catch/condition Or find this in submenu
3) If you just placed cursor it will find surrounding statements around cursor position
 else if you selected code then it will check weather these are statements or not
5) If statements -> It will wrap that code in TryCatch/Condition
 else error (Please select valid statements)

`AppInit.appReady(function () {
            brackets.test.doneLoading = true;
        });`

Select whole code or just place cursor on 3rd line it will wrap whole code in try catch block/Condition block

`try {
            AppInit.appReady(function () {
                brackets.test.doneLoading = true;
            });
        } catch (e) {
            console.log(e.message);
        }`
Put selection on 2nd line and do wrap in try catch/condition it will wrap that line only.

Know Issue:- Some time indention goes wrong in case of wrap in condition

**Convert to arrow function**
1) Put cursor anywhere in anonymous function and click convert to arrow function
If one param structure will be `param => {statements}`
If zero or more than one param `(param1, param2) => {statements}` 
In case of one statement, it will remove "{, }" and ";"as well

**Create Setters and Getters for given property**
1) Put cursor on property
2) Create getter/setter
`get getTest() {
 return this.test;
},
set setTest(val) {
this.test = val;
}`

Please review @nethip @boopeshmahendran @petetnt @swmitra @ficristo @saurabh95

### CODE REVIEW

**Comment by nethip** [comment]
Is it OK to squash all of your commits into one single commit?

File: src/extensions/default/JavaScriptRefactoring/RefactoringUtils.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    var _ = brackets.getModule("thirdparty/lodash");
+
+    var AcornLoose           = brackets.getModule("thirdparty/acorn/dist/acorn_loose"),
+        ASTWalker            = brackets.getModule("thirdparty/acorn/dist/walk");
+
+    var templates = JSON.parse(require("text!Templates.json"));
+
+    /**
+     * Note - Don't use this if you are in a batch operation and you already made some changes
```
Comment (nethip) [inline src/extensions/default/JavaScriptRefactoring/RefactoringUtils.js]: This entire comment section needs to be rewritten.

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        return {promise: ternPromise};
+    }
+
+    //Do rename of identifier which is at cursor
+    function handleRename() {
+        var editor = EditorManager.getActiveEditor(),
+            offset, handleFindRefs;
+
+        if (editor.getSelections().length > 1) {
+            editor.displayErrorMessageAtCursor("Rename doesn't work in case of multicursor");
```
Comment (nethip) [inline src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: This string needs to be externalized. And correct the wording as well.

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            };
+
+        var ternPromise = getRefs(fileInfo, offset);
+
+        return {promise: ternPromise};
+    }
+
+    //Do rename of identifier which is at cursor
+    function handleRename() {
+        var editor = EditorManager.getActiveEditor(),
```
Comment (nethip) [inline src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: Are you going to allow rename inside quick edit session?

File: src/extensions/default/JavaScriptRefactoring/WrapSelection.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    //Convert function to arrow function
+    function convertToArrowFunction() {
+        initializeRefactoringSession();
+        //Handle when there is no selected line
+        var funcExprNode = current.findSurroundASTNode(current.ast, {start: current.startIndex}, ["FunctionExpression"]);
+
+        if (!funcExprNode || funcExprNode.type !== "FunctionExpression" || funcExprNode.id) {
+            current.editor.displayErrorMessageAtCursor("Cursor is not inside function expression");
```
Comment (nethip) [inline src/extensions/default/JavaScriptRefactoring/WrapSelection.js]: String needs to be externalized.

**Comment by nethip** [comment]
@navch  All error popup strings need to go into `Strings.js` file. Please make this change. Also consider re-wording those strings.

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js (hunk 3 above)
Comment (navch) [reply src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: Yes. Rename is working great in quick edit as well for local variables. Renaming function calls could be a problem. According to me we should allow rename in quick edit as well. Please let me know if want me to disable rename in quick edit.

**Comment by navch** [comment]
Thanks @nethip for reviewing this. I addressed all review comments.

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                });
+            }
+        }
+
+        /**
+         * Check if references are in this file only
+         * If yes then select all references
+         */
+        handleFindRefs = function (refsResp) {
+            function isInSameFile(obj) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: `isInSameFile` could be externalized from the function

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                    result.reject();
+                });
+            }
+        }
+
+        /**
+         * Check if references are in this file only
+         * If yes then select all references
+         */
+        handleFindRefs = function (refsResp) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: Maybe cleaner if this one was assigned inline above `requestFindReferences`?

`function handleFindRefs(refsResp)...`

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        /**
+         * Make a find ref request.
+         * @param {Session} session - the session
+         * @param {number} offset - the offset of where to jump from
+         */
+        function requestFindReferences(session, offset) {
+            var response = requestFindRefs(session, session.editor.document, offset);
+
+            if (response || response.hasOwnProperty("promise")) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: This should probably be `if (response && response.hasOwnProperty("promise")`?

File: src/nls/root/strings.js [hunk 8]
```diff
@@ -674,6 +674,20 @@ define({
...
     "DETECTED_EXCLUSION_INFO"                   : "{APP_NAME} ran into trouble processing <span class='dialog-filename'>{0}</span>.<br><br>This file will no longer be processed for code hints, Jump to Definition or Quick Edit. To re-enable this file, open <code>.brackets.json</code> in your project and edit <code>jscodehints.detectedExclusions</code>.<br><br>This is likely a {APP_NAME} bug. If you can provide a copy of this file, please <a href='https://github.com/adobe/brackets/wiki/How-to-Report-an-Issue'>file a bug</a> with a link to the file named here.",
 
+
+    // extensions/default/JavaScriptRefactoring
+    "CMD_REFACTORING_RENAME"                    : "Rename",
+    "CMD_REFACTORING_TRY_CATCH"                 : "Wrap in Try Catch",
+    "CMD_REFACTORING_CONDITION"                 : "Wrap in Condition",
+    "CMD_REFACTORING_GETTERS_SETTERS"           : "Create Getters Setters",
+    "CMD_REFACTORING_ARROW_FUNCTION"            : "Convert to Arrow Function",
+    "DESCRIPTION_CODE_REFACTORIG"               : "Enable/disable JavaScript Code Refactoring",
```
Comment (petetnt) [inline src/nls/root/strings.js]: DESCRIPTION_CODE_REFACTORING

File: src/extensions/default/JavaScriptRefactoring/main.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        Strings              = brackets.getModule("strings"),
+        RenameIdentifier     = require("RenameIdentifier"),
+        WrapSelection        = require("WrapSelection");
+
+    var jsRefactoringEnabled     = true;
+
+
+    // This preference controls whether to create a session and process all JS files or not.
+    PreferencesManager.definePreference("refactoring.JSRefactoring", "boolean", true, {
+        description: Strings.DESCRIPTION_CODE_REFACTORIG
```
Comment (petetnt) [inline src/extensions/default/JavaScriptRefactoring/main.js]: DESCRIPTION_CODE_REFACTORING

**Review by petetnt — COMMENTED** [review]
Few review comments

File: src/JSUtils/node/TernNodeDomain.js [hunk 10]
```diff
@@ -236,6 +236,44 @@ function buildRequest(fileInfo, query, offset) {
...
+ * @param {{type: string, name: string, offsetLines: number, text: string}} fileInfo
+ * - type of update, name of file, and the text of the update.
+ * For "full" updates, the whole text of the file is present. For "part" updates,
+ * the changed portion of the text. For "empty" updates, the file has not been modified
+ * and the text is empty.
+ * @param {{line: number, ch: number}} offset - the offset into the
+ * file for cursor
+ */
+ function getRefs(fileInfo, offset) {
+    var request = buildRequest(fileInfo, "refs", offset);
```
[the remark this answers is not in the document]
Comment (navch) [reply src/JSUtils/node/TernNodeDomain.js]: This should not be the case. Anyways I handled this when we are creating fileInfo.

**Comment by navch** [comment]
Thanks @petetnt for reviewing this. I addressed all review comments.

File: src/JSUtils/node/TernNodeDomain.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (petetnt) [reply src/JSUtils/node/TernNodeDomain.js]: 👍

File: src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js (hunk 3 above)
Comment (nethip) [reply src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: OK. I think it is better to disable rename for quick edit. There are chances that once you invoke a quick edit on a function name, we only show the contents of the function. But there may be variables whose scope is outside the function and I am not sure how you are going to show the feedback for renaming variable references which are not shown inside quick edit.

Comment (nethip) [reply src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: Let us do this. I will merge this PR. But please raise another PR addressing this concern.

Comment (navch) [reply src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js]: Sure @nethip I will do that. Thanks for suggestion.

**Pull request merged**